python/cudf/cudf/core/dataframe.py

```diff
@@ -758,77 +777,121 @@ def __init__(
+                )
+                for col_label in columns
+            }
+        elif is_list_like(data):
+            super().__init__()
+            if len(data) > 0 and is_scalar(data[0]):
+                if columns is not None:
+                    data = dict(zip(columns, [data]))
+                    rangeindex = isinstance(
+                        columns, (range, pd.RangeIndex, cudf.RangeIndex)
                     )
                 else:
-                    self._init_from_list_like(
-                        data, index=index, columns=columns
-                    )
-                self._check_data_index_length_match()
+                    data = dict(enumerate([data]))
+                    rangeindex = True
+                new_df = DataFrame(data=data, index=index)
+
+                self._data = new_df._data
+                self._index = new_df._index
+                self._data._level_names = (
+                    tuple(columns.names)
+                    if isinstance(columns, pd.Index)
+                    else self._data._level_names
+                )
+                self._data.rangeindex = rangeindex
+            elif len(data) > 0 and isinstance(data[0], Series):
+                self._init_from_series_list(
+                    data=data, columns=columns, index=index
+                )
             else:
-                if not is_dict_like(data):
-                    raise TypeError("data must be list or dict-like")
+                self._init_from_list_like(data, index=index, columns=columns)
+            self._check_data_index_length_match()
 
-                self._init_from_dict_like(
-                    data, index=index, columns=columns, nan_as_null=nan_as_null
-                )
-                self._check_data_index_length_match()
+            if dtype:
+                self._data = self.astype(dtype)._data
+            return
+        elif is_dict_like(data):
+            result = self._init_from_dict_like(
+                data, index=index, nan_as_null=nan_as_null
+            )
+            col_dict = result[0]
+            index = result[1]
+            columns, columns_from_data = result[2], columns
+            col_is_multiindex = isinstance(columns, pd.MultiIndex)
+        else:
+            raise TypeError(
+                f"data must be list or dict-like, not {type(data).__name__}"
+            )
+        col_accessor = ColumnAccessor(
+            col_dict,
+            multiindex=col_is_multiindex,
+            rangeindex=col_is_rangeindex,
+            label_dtype=col_dtype,
+        )
+        super().__init__(col_accessor, index=index)
+        self._check_data_index_length_match()
+        if columns_from_data is not None:
+            # TODO: This there a better way to do this?
+            columns_from_data = as_index(columns_from_data)
+            reindexed = self.reindex(
+                columns=columns_from_data.to_pandas(), copy=False
+            )
```

Comment:
**question** What are you trying to do conceptually that you would like a better way for?